### PR TITLE
Inverse print direction for holes

### LINKED ISSUE #11630: wrong direction of outer walls on the inside of a hollow part
### Is there an existing issue for this problem?

- [x] I have searched the existing issues

### OrcaSlicer Version

all

### Operating System (OS)

Linux

### OS Version

not relevant

### Additional system information

irelevant

### Printer

irelevant

### How to reproduce

For some time now I've been wondering if the direction of walls located on the inside of e.g. a mug should be printed in the opposite direction compared to those on the outside.

And yesterday I finally stumbled over situation that proved that my skepticism was not unreasonable.

### Actual results

Here is the situation:

1. wall direction on the outside (ccw)

Image

<br>
<br>

2. wall direction on the inside (ccw)

Image

<br>
<br>

3. wall direction on the inside (cw)

Image

<br>
<br>

4. wall direction on the outside (ccw)

Image

In this case the direction of movement on walls located on the outside of a part is consistent whereas the direction of the movement on walls located on the inside of a hollow part changes depending on the whether there are holes in the wall or not.

### Expected results

So, in order to make the direction of walls be consistent over the whole part the default direction of movement on walls located on the inside of a hollow part need to be opposite to the direction of movement on walls located on the corresponding outside of a part.

### Project file & Debug log uploads

-

### Checklist of files to include

- [ ] Log file
- [ ] Project file

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by RF47**
# Description

This pull request is based on an idea from @discip, reversing the printing direction of the holes to improve printing consistency on internal walls (see #11630).
We also incorporated the changes made by @vovodroid in #6193.

## Requisites :
merge #6193

# Screenshots/Recordings/Graphs

Main idea:
red CCW, blue CW

## Before / After:
image 
image

## Before:
image

## After:
image

Close #11630

### CODE REVIEW

**Comment by igiannakas** [comment]
Awesome!

one thing to check is how profiles that use a saved auto direction handle this config change - where the profile doesn’t have that value available in the config.

**Comment by RF47** [comment]
> Awesome!
> 
> one thing to check is how profiles that use a saved auto direction handle this config change - where the profile doesn’t have that value available in the config.

I'm not sure if this handling is necessary. If the profile was set to “auto”, it switches to the default mode, which is CCW.

File: src/libslic3r/PrintConfig.cpp [hunk 1]
```diff
@@ -2013,16 +2012,14 @@ void PrintConfigDef::init_fff_params()
     def = this->add("wall_direction", coEnum);
     def->label = L("Wall loop direction");
     def->category = L("Quality");
-    def->tooltip = L("The direction which the wall loops are extruded when looking down from the top.\n\nBy default all walls are extruded in counter-clockwise, unless Reverse on even is enabled. Set this to any option other than Auto will force the wall direction regardless of the Reverse on even.\n\nThis option will be disabled if spiral vase mode is enabled.");
+    def->tooltip = L("The direction which the contour wall loops are extruded when looking down from the top.\nHoles are printed in the opposite direction to the contour.  \n\nBy default all contours are extruded in counter-clockwise, unless Reverse on even is enabled. Set this to any option other than Auto will force the wall direction regardless of the Reverse on even.\n\nThis option will be disabled if spiral vase mode is enabled.");
```
Comment (ianalexis) [inline src/libslic3r/PrintConfig.cpp]: If the #6193 is merged we have to clean the "Auto" and "Default" part of the description and we can reuse that part to explain a little why internal are reversed.
```suggestion
    def->tooltip = L("The direction which the contour wall loops are extruded when looking down from the top.\nHoles are printed in the opposite direction to the contour to maintain alignment with layers whose contour polygons are incomplete and change direction, also partially forming the contour of a hole..  \n\nThis option will be disabled if spiral vase mode is enabled.");
```

File: src/libslic3r/PerimeterGenerator.cpp [hunk 2]
```diff
@@ -511,7 +514,13 @@ static ExtrusionEntityCollection traverse_extrusions(const PerimeterGenerator& p
...
             if (extrusion->is_closed) {
                 ExtrusionLoop extrusion_loop(std::move(paths), pg_extrusion.is_contour ? elrDefault : elrHole);
-                extrusion_loop.make_counter_clockwise();
+                if (perimeter_generator.config->wall_direction == WallDirection::CounterClockwise)
+                    extrusion_loop.make_counter_clockwise();
+                else
+                    extrusion_loop.make_clockwise();
+                if (!pg_extrusion.is_contour){
+                    extrusion_loop.reverse();
+                }
```
Comment (ianalexis) [inline src/libslic3r/PerimeterGenerator.cpp]: It can be simplified and avoid a reverse step with a XNOR.
```suggestion
                if ((perimeter_generator.config->wall_direction == WallDirection::CounterClockwise) == pg_extrusion.is_contour)
                    extrusion_loop.make_counter_clockwise();
                else
                    extrusion_loop.make_clockwise();
```

**Review by ianalexis — CHANGES_REQUESTED** [review]

File: src/libslic3r/PerimeterGenerator.cpp (hunk 2 above)
Comment (ianalexis) [reply src/libslic3r/PerimeterGenerator.cpp]: Will push an update because we missed classic wall generator

**Comment by valerii-bokhan** [comment]
Hey, I'm sorry for writing before checking the code. Am I correct in saying that this logging will be somewhat ignored if the `Reverse on Even` (`overhang_reverse`) is enabled?

If I enable `Reverse on Even` for each layer (threshold = 0), how will it work?

**Comment by ianalexis** [comment]
> If I enable `Reverse on Even` for each layer (threshold = 0), how will it work?

Reverse on even off 
Mode : counter clock

- any layer
    - contour = counter clock
    - hole = clock

Mode : counter clock

- Layer uneven
    - contour = counter clock
    - hole = clock
- Layer even
    - contour =clock 
    - hole = counter clock

The rest of the behavior it's the same, we only reverse the hole to match the contour because if we not when the contour for some reason as any gap in the wall meet a perimeter matching a hole part it was creating the same artifact as it does while using reverse on even but just on the transition part.
If you use reverse on even it will reverse to its opposite.

**Comment by RF47** [comment]
If you use reverse on even including the outer perimeters, in that case it is impossible to achieve consistency between layers.

**Comment by discip** [comment]
Great!
Thank you!
Working on windows.
Need to test on Linux once I figured out what's wrong with Flatpak on my system.

**Comment by RF47** [comment]
> @RF47 Confirmed to run both on win10 and also on Linux (Manjaro GNOME)!

Thank you for your tests. 
I think this is a good idea and improves the quality of the internal perimeters.

**Comment by igiannakas** [comment]
Merged in by dev branch for long term testing. Awesome work!

**Comment by igiannakas** [comment]
Have done some initial tests. For single color prints this is exceptional.

However for multi color / multi material we need to do some more testing. Why?

image

image

In the above scenario the inner circle (blue) will have its **outer perimeter printed counter clock wise (good)**. However the **inner perimeter of the yellow ring will be printed clock wise.** 

What this will result is a **slight misalignment between the two parts** as they have **neighbouring walls that are printed with reversing wall directions.** 

Will this be a problem? Not sure. It needs testing (**multi color / material prints with concentric geometries**).

In addition, what this change will result is a reversing direction of printing in small features, like the bench flag pole. This has the potential to lead to slightly more prominent seams as we are backing on ourselves and printing next to a partially hot perimeter still.

image

We have noticed a similar artefact when reverse on even was enabled with the outer perimeters set to not reverse. there on every other layer the inner perimeter was printed in the opposite direction, causing the nozzle to print back over hot plastic.

All in all, while this PR fixes one issue, it needs more testing to ensure it does not generate another under the specific below two cases:
1. Concentric multi-material features
2. A hole with 2 wall size to the next contour.

Thoughts welcome!!

**Comment by igiannakas** [comment]
MMU project file here: 
[EMU Demo Project 0.4 Nozzle.3mf.zip](https://github.com/user-attachments/files/25881666/EMU.Demo.Project.0.4.Nozzle.3mf.zip)

**Comment by RF47** [comment]
I don't know if there is an optimal strategy that covers all possible cases, but we can always make this behaviour optional if necessary.

**Comment by RF47** [comment]
I made a simple model for testing purposes and confirmed the problem with only two perimeters. In such cases, it may be best to maintain the direction of rotation or consider using staggered seams.
I cannot test the multi-color issue on my machine, so print tests would need to be carried out to see how much it affects the result.
image
[test wall direction.zip](https://github.com/user-attachments/files/25884383/test.wall.direction.zip)

**Comment by RF47** [comment]
> 2\. A hole with 2 wall size to the next contour.

Update: I added logic to maintain the direction of rotation in these cases.

File: src/libslic3r/PerimeterGenerator.cpp [hunk 3]
```diff
@@ -518,6 +518,18 @@ static ExtrusionEntityCollection traverse_extrusions(const PerimeterGenerator& p
                     extrusion_loop.make_counter_clockwise();
                 else
                     extrusion_loop.make_clockwise();
+                // Detect thin wall holes, to prevent drag hot plastic while changing direction. We define thin wall hole as a hole with
+                // only one perimeter.
+                std::unordered_set<int> perimeter_counts;
+                for (const auto& pg_ext : pg_extrusions) {
+                    if (pg_ext.extrusion && !pg_ext.extrusion->empty()) {
+                        perimeter_counts.insert(pg_ext.extrusion->inset_idx);
```
Comment (igiannakas) [inline src/libslic3r/PerimeterGenerator.cpp]: Im a bit confused with this. We are creating a new set and are inserting the inset index of the extrusion. But then later we are not checking that but rather the size of the set. 

Isn't that the same as doing a pg_extrusions.size() and skipping the for loop?

Comment (RF47) [reply src/libslic3r/PerimeterGenerator.cpp]: Maybe so; perhaps I should check the size equal to 2 (one perimeter and one hole). I'll give it a try later.

Comment (RF47) [reply src/libslic3r/PerimeterGenerator.cpp]: good idea, much simpler

**Comment by Felix14-v2** [comment]
> > 2. A hole with 2 wall size to the next contour.
> 
> Update: I added logic to maintain the direction of rotation in these cases.

Sounds promising. A bit off-topic, but I think this logic can be useful to fix another issue with such thin walls:

https://github.com/user-attachments/assets/47f0de16-404e-48ac-9272-611a5a18d19f

Both perimeters at the top are considered as external, and the printing order is not maintained there. Could this logic be useful to handle this case as well?
3MF: [Thin walls.zip](https://github.com/user-attachments/files/25994274/Thin.walls.zip)

**Comment by valerii-bokhan** [comment]
@Felix14-v2, hey, can you re-upload the video, please? It's not available for some reason. I tried to watch it on both the PC and mobile, but it's unavailable.

**Comment by Felix14-v2** [comment]
I think re-uploading wouldn't help. The video is encoded with H.265 and plays fine in my desktop and mobile browsers (Yandex and Firefox Mobile). I'll try to convert it to H.264

**Comment by Felix14-v2** [comment]
Try this one

https://github.com/user-attachments/assets/d2f81e4d-2ce0-4db2-8901-434a4f305cee

**Comment by RF47** [comment]
> > > 2. A hole with 2 wall size to the next contour.
> > 
> > 
> > Update: I added logic to maintain the direction of rotation in these cases.
> 
> Sounds promising. A bit off-topic, but I think this logic can be useful to fix another issue with such thin walls:
> 
>  2026.March.14.12.54.mp4 
> Both perimeters at the top are considered as external, and the printing order is not maintained there. Could this logic be useful to handle this case as well? 3MF: [Thin walls.zip](https://github.com/user-attachments/files/25994274/Thin.walls.zip)

This problem can be avoided by using aligned back seams

**Comment by Felix14-v2** [comment]
> This problem can be avoided by using aligned back seams

I don't mean seams (we already have the seam painting tool anyway), I mean the perimeter print order (please, watch the video)

**Comment by RF47** [comment]
> > This problem can be avoided by using aligned back seams
> 
> I don't mean seams (we already have the seam painting anyway), I mean the perimeter print order (please, watch the video)

I think is fixed now

**Comment by Felix14-v2** [comment]
Now the print order is always inner/outer:

https://github.com/user-attachments/assets/c9e2954c-bc89-491c-9acc-d7834a190ab4

If fixing that is out of scope of this PR, I'll create a new issue for this behaviour. I just thought that the root of the problem with wall order may be related to the same code.

**Comment by RF47** [comment]
> Now the print order is always inner/outer:

The problem only occurred with Arachne; I think I've sorted it out.
Could you give it another try?

**Comment by Felix14-v2** [comment]
Works fine now with Arachne!

Found another edge-case though (I don't know is it related or not) with 3 perimeters (2 outer and 1 inner)

https://github.com/user-attachments/assets/668cbb4f-185b-47c7-82a4-366fd5ef129a

**Comment by RF47** [comment]
> Works fine now with Arachne!
> 
> Found another edge-case though (I don't know is it related or not) with 3 perimeters (2 outer and 1 inner)
> 
>  2026.March.15.00.08.mp4

I’m not addressing that case in this PR; I’m only focusing on the two-wall condition. I think that, to avoid complicating this PR further, it’s best to resolve it in a separate issue – I want to keep this as simple as possible.

File: src/libslic3r/PerimeterGenerator.cpp [hunk 4]
```diff
@@ -1414,18 +1433,17 @@ void PerimeterGenerator::process_classic()
...
-            const WallDirection wall_direction = config->wall_direction;
-            if (wall_direction != WallDirection::Auto) {
-                // Skip steep overhang detection if wall direction is specified
+            if (!config->overhang_reverse) {
+                // Skip steep overhang detection no reverse is specified
                 steep_overhang_contour = true;
                 steep_overhang_hole    = true;
             }
-            ExtrusionEntityCollection entities = traverse_loops(*this, contours.front(), thin_walls, steep_overhang_contour, steep_overhang_hole);
+			ExtrusionEntityCollection entities = traverse_loops(*this, contours.front(), thin_walls, steep_overhang_contour, steep_overhang_hole, false);
```
Comment (ianalexis) [inline src/libslic3r/PerimeterGenerator.cpp]:
```suggestion
            ExtrusionEntityCollection entities = traverse_loops(*this, contours.front(), thin_walls, steep_overhang_contour, steep_overhang_hole, false);
```

**Comment by SoftFever** [comment]
@RF47  
Thank you for your excellent work!  
PR #6193 has been merged.  
Could you please review the conflicts?

**Comment by RF47** [comment]
@SoftFever, regarding this PR, we need to run some multicolor print tests just to make sure it doesn't affect the quality at the color boundaries; I don't have a multicolor printer to test that.

**Comment by RF47** [comment]
> @RF47 Thank you for your excellent work! PR #6193 has been merged. Could you please review the conflicts?

done

**Comment by SoftFever** [comment]
> @SoftFever, regarding this PR, we need to run some multicolor print tests just to make sure it doesn't affect the quality at the color boundaries; I don't have a multicolor printer to test that.

okay

**Comment by igiannakas** [comment]
I’ve been testing this Pr with multi color prints and regular prints. Nothing weird to report.

**Comment by igiannakas** [comment]
IMG_0652
IMG_0500
IMG_0406
IMG_0374
IMG_0353

**Comment by RF47** [comment]
nice

**Comment by SoftFever** [comment]
> I’ve been testing this Pr with multi color prints and regular prints. Nothing weird to report.

You're a legend!

**Comment by discip** [comment]
Is the behavior for the first layer intentionally kept? 😊

For the first layer the print direction for holes is not inversed.

**Comment by RF47** [comment]
> Is the behavior for the first layer intentionally kept? 😊
> 
> For the first layer the print direction for holes is not inversed.

It’s not intentional; in the tests I carried out, the rotation for the holes in the first layer is reversed.

**Comment by discip** [comment]
[sleeve.3mf.txt](https://github.com/user-attachments/files/26680768/sleeve.3mf.txt)

||inner|dir|outer|dir
:-:|:-:|:-|:-:|:-
1st layer|image|CW|image|CW
2nd layer|image|CCW|image|CW

**Comment by RF47** [comment]
It’s an edge case; it happens because there are only two perimeters and it’s interpreted as a thin_wall_hole. This is because the ‘only one wall on first layer’ option is enabled.

**Comment by discip** [comment]
But it shouldn't be to hard to cover this case as well., right? 😊

**Comment by RF47** [comment]
> But it shouldn't be to hard to cover this case as well., right? 😊

I don't think the risk of breaking something else justifies fixing this, as it's harmless.

**Comment by discip** [comment]
Fair enough.
